crates/gc/src/lib.rs

```diff
@@ -794,3 +796,128 @@ impl<'a> RemapContext<'a> {
         }
     }
 }
+
+fn gc_locals(module: &mut Module) {
+    let mut code = None;
+    let mut types = None;
+    let mut functions = None;
+    let mut locals = None;
+    let mut imported_functions = 0;
+    for section in module.sections_mut() {
+        match section {
+            Section::Import(s) => imported_functions = s.functions(),
+            Section::Code(s) => code = Some(s),
+            Section::Function(s) => functions = Some(s),
+            Section::Type(s) => types = Some(s),
+            Section::Name(NameSection::Local(s)) => locals = Some(s),
+            _ => {}
+        }
+    }
+    let code = match code {
+        Some(s) => s,
+        None => return,
+    };
+    let types = types.unwrap();
+    let functions = functions.unwrap();
+
+    for (i, body) in code.bodies_mut().iter_mut().enumerate() {
+        let ty_idx = functions.entries()[i].type_ref();
+        let ty = match &types.types()[ty_idx as usize] {
+            Type::Function(f) => f,
+        };
+        gc_body((i + imported_functions) as u32, body, ty, &mut locals);
+    }
+}
+
+/// Each body of a function has a list of locals, and each local has a type and
+/// a number of how many locals of this type there are. The purpose of this is
+/// to collapse something like:
+///
+///     Local::new(1, ValueType::I32);
+///     Local::new(1, ValueType::I32);
+///
+/// to ...
+///
+///     Local::new(2, ValueType::I32);
+///
+/// to encode this a bit more compactly
+fn gc_body(
+    idx: u32,
+    body: &mut FuncBody,
+    ty: &FunctionType,
+    names: &mut Option<&mut LocalNameSection>,
+) {
+    let mut local_tys = ty.params().to_vec();
+    for local in body.locals_mut().drain(..) {
+        for _ in 0..local.count() {
+            local_tys.push(local.value_type());
+        }
+    }
+    let mut used = vec![false; local_tys.len()];
+
+    for instr in body.code().elements() {
+        match instr {
+            Instruction::GetLocal(i) => used[*i as usize] = true,
+            Instruction::SetLocal(i) => used[*i as usize] = true,
+            Instruction::TeeLocal(i) => used[*i as usize] = true,
+            _ => {}
+        }
+    }
+
+    let mut map = vec![u32::max_value(); local_tys.len()];
```

Comment:
Why not just use `Option<u32>` and `unwrap` down below instead of asserting against `u32::max_value()`?

Reply:
👍